File: src/cpp/multithreading/02_thread_subclass.cpp
#include <chrono>
#include <functional>
#include <iostream>
#include <thread>

class MyThread {
public:
  MyThread(std::function<void()> function)
      : function_(function), thread_(&MyThread::run, this) {}

  ~MyThread() { thread_.join(); }

  void run() {
    std::cout << "Function started" << std::endl;
    function_();
    std::cout << "Function finished" << std::endl;
  }

private:
  std::function<void()> function_;
  std::thread thread_;
};

void my_function() {
  std::cout << "my_function goes to sleep" << std::endl;
  std::this_thread::sleep_for(std::chrono::seconds(5));
  std::cout << "my_function wakes up" << std::endl;
}

int main() {
  // Start threads with different functions
  MyThread thread_a(my_function);
  MyThread thread_b(
      []() { std::cout << "it's me, the lambda function" << std::endl; });

  // Threads are joined automatically in MyThread's destructor

  std::cout << "Main thread finished" << std::endl;

  return 0;
}

File: src/python/multithreading/02_thread_subclass.py
"""
This module demonstrates the usage of custom threads in Python by implementing a
`MyThread` class, which is a subclass of the built-in `Thread` class.

The `MyThread` class is initialized with a function and automatically starts the
thread upon instantiation. It then runs the given function and prints the start
and finish status of the function.

Two threads are created in the `main()` function, one with a custom function called
`my_function` and another with a lambda function. Both threads are joined, and the
main thread waits for them to finish before terminating.
"""

import time
from threading import Thread


class MyThread(Thread):
    def __init__(self, function, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.function = function
        self.start()

    def run(self):
        print(f"{self.function.__name__} started")
        self.function()
        print(f"{self.function.__name__} finished")


def my_function():
    print("my_function goes to sleep")
    time.sleep(5)
    print("my_function wakes up")


def main():
    # Start threads with different functions
    thread_a = MyThread(my_function)
    thread_b = MyThread(lambda: print("it's me, the lambda function"))

    # Wait for both threads to finish
    for thread in (thread_a, thread_b):
        thread.join()

    print("Main thread finished")


if __name__ == "__main__":
    main()
